Dev

Flodder23 opened the pull request:
The other request got a bit long, so I made this one to make it easier

Commit:
Introduced the FPS slider

Commit:
Implimented Generation counter

Also a few PEP8 fixes

Commit:
Added a "Not limited by FPS" option

Commit:
Better personalisation in config, "Wrap" option for borders

Added options for min and max value on FPS slider, added "Speed" label to slider, added option to have a "Wrap" instead of a cushion to combat the problem of edges. Also added a random tile pattern when game loads.

Commit:
changed FPS to GPS, improved some varibale names

Commit:
renamed some variables

also moved update_board and draw_board out of the main loop

Commit:
Update .gitignore

Commit:
Update .gitignore

Commit:
made Board a class

Also made GameState and Widgets classes

Commit:
updated preset file

Commit:
Fixed FPS Slider issue

max-kov commented on config.py:
```diff
@@ -1,10 +1,86 @@
+import pygame
+
 Dead = 0
 Square = 1
```
I've used [enums](https://docs.python.org/3/library/enum.html) in my pool game for striped balls and current player values. You can check it out, It does make the code more readable, but it's not an easy task to implement.[ Here's an example from my code](https://github.com/max-kov/pool/blob/f0a03fb9e928f7cfa247ba5e9a8c3944148b1a31/pool/gamestate.py#L201)

max-kov commented on config.py:
```diff
@@ -1,10 +1,86 @@
+        self.NextState = state
+
+
+class Board:
```
Classes shouldn't be defined here, this file is for constants only. I'd make a board.py file with that class

joegibby replied:
done

max-kov commented on config.py:
```diff
@@ -1,10 +1,86 @@
+        self.SliderY = B.Size * B.Width + B.Edge / 2 + self.ButtonSize / 2
+
+
+def write(screen, x, y, text, colour, size, rotate=0, alignment=("left", "top")):
```
Again, this file is for constant variables, so that the user could go into this file and change stuff. The user probably won't want to change the function itself. You can make an `alignment` variable here, and then use it in the main game file.

joegibby replied:
done

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
-    for a in range(config.Cushion, config.Cushion + config.Width):
-        for b in range(config.Cushion, config.Cushion + config.Height):
-            if Board[a][b].NextState == config.Dead:
+    pygame.display.set_caption("Game of Life - Generation " + str(board.Generations))
```
Function `draw_board` shouldn't set the caption of the window.

joegibby replied:
that is now done when the board goes forward a generation (in `take_turn` function)

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
-            range(config.Width + 2 * config.Cushion)]
-
-
 def draw(state, a, b, colour):
```
rename to `draw_cell()`

joegibby replied:
it is in the cell class now

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
+    s = Board.Size - Board.Edge
     pygame.draw.rect(Screen, (255, 255, 255), (x, y, s, s))
     if state == config.Square:
         pygame.draw.rect(Screen, colour, (x, y, s, s))
```
This is very confusing. The function is drawing the cell not at `a` but at `(a - Board.Cushion) * Board.Size + Board.Edge / 2` and then draws the second point at `Board.Size - Board.Edge` which sounds like the edge of the window. And then if the state `config.Square` it draws the cell a second time, but this time in colour `colour`. Not saying that this must or can be changed, but I'd look into simplifying this. Drawing a cell at (x,y) shouldn't be this hard

joegibby replied:
a and b are the places in the board, x and y are the coordinates on the screen

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
+                board = update_board(board)
+                draw_board(board)
+        if pygame.key.get_pressed()[pygame.K_f]:
+            game_state.GPSLimit = not game_state.GPSLimit
```
`GPSLimit` sounds like an integer value, i'd rename to something like `IsGpsLimited`

joegibby replied:
renamed to GPSIsLimited

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
         if pygame.key.get_pressed()[pygame.K_RETURN]:
-            board = clean_board()
+            board = config.Board()
+            board = update_board(board)
```
You create a `board` and then immediately overwrite it with `update_board`?

joegibby replied:
done

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
-            board[a][b].kill()
-    return board, paused, one_turn
+            board.Cell[a][b].kill()
+            board = update_board(board)
```
`Update_board` sounds like like it should be a part of the `Board` Class

joegibby replied:
done

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
 
 
-def draw_board():
+def draw_board(board):
```
everything board related should go in the board class

joegibby replied:
done

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
+    pygame.display.update()
 
 
 def take_turn(board):
```
maybe `do_turn` or `next_turn`

max-kov commented on main.py:
```diff
@@ -41,99 +40,175 @@ def check(a, b):
-Board = clean_board()
-Paused = True
+pygame.event.set_allowed(None)
+pygame.event.set_allowed(pygame.MOUSEMOTION)
```
Cut the copypasta in a separate file/function

joegibby replied:
made it into a list

max-kov commented on preset.py:
```diff
@@ -3,144 +3,144 @@
+        board.Cell[a + 2][b].kill()
+        board.Cell[a][b + 1].kill()
+        board.Cell[a + 1][b + 1].kill()
+        board.Cell[a + 2][b + 1].birth(config.Square)
```
Make a function which draws a preset from a 2d Boolean matrix, and then fill your config file with presets. For example, if you pass 
```
[[True,False,False],
[False,True,False],
[False,False,True]]
```
in the function you would get
```
x__
_x_
__x
```

joegibby replied:
done

Commit:
Broken code

Overhauled some classes, code isn't working atm, just uploading to save progress so far

Commit:
Works now`

improved some code

Commit:
Fixed presets+wrap issue

Commit:
Moved write function

Commit:
Moved the changing of the display caption

Commit:
Tidied up repeated lines
